File: sdk/v2/core/jobs.go
package core

import (
	"context"
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"time"

	rm "github.com/brigadecore/brigade/sdk/v2/internal/restmachinery"
	"github.com/brigadecore/brigade/sdk/v2/meta"
	"github.com/brigadecore/brigade/sdk/v2/restmachinery"
)

// JobPhase represents where a Job is within its lifecycle.
type JobPhase string

const (
	// JobPhaseAborted represents the state wherein a Job was forcefully
	// stopped during execution.
	JobPhaseAborted JobPhase = "ABORTED"
	// JobPhaseFailed represents the state wherein a Job has run to
	// completion but experienced errors.
	JobPhaseFailed JobPhase = "FAILED"
	// JobPhasePending represents the state wherein a Job is awaiting
	// execution.
	JobPhasePending JobPhase = "PENDING"
	// JobPhaseRunning represents the state wherein a Job is currently
	// being executed.
	JobPhaseRunning JobPhase = "RUNNING"
	// JobPhaseSchedulingFailed represents the state wherein a job was not
	// scheduled due to some unexpected and unrecoverable error encountered by the
	// scheduler.
	JobPhaseSchedulingFailed JobPhase = "SCHEDULING_FAILED"
	// JobPhaseSucceeded represents the state where a Job has run to
	// completion without error.
	JobPhaseSucceeded JobPhase = "SUCCEEDED"
	// JobPhaseTimedOut represents the state wherein a Job has has not completed
	// within a designated timeframe.
	JobPhaseTimedOut JobPhase = "TIMED_OUT"
	// JobPhaseUnknown represents the state wherein a Job's state is unknown. Note
	// that this is possible if and only if the underlying Job execution substrate
	// (Kubernetes), for some unanticipated reason, does not know the Job's
	// (Pod's) state.
	JobPhaseUnknown JobPhase = "UNKNOWN"
)

// Job represents a component spawned by a Worker to complete a single task
// in the course of handling an Event.
type Job struct {
	// Spec is the technical blueprint for the Job.
	Spec JobSpec `json:"spec"`
	// Status contains details of the Job's current state.
	Status *JobStatus `json:"status"`
}

// MarshalJSON amends Job instances with type metadata so that clients do not
// need to be concerned with the tedium of doing so.
func (j Job) MarshalJSON() ([]byte, error) {
	type Alias Job
	return json.Marshal(
		struct {
			meta.TypeMeta `json:",inline"`
			Alias         `json:",inline"`
		}{
			TypeMeta: meta.TypeMeta{
				APIVersion: meta.APIVersion,
				Kind:       "Job",
			},
			Alias: (Alias)(j),
		},
	)
}

// JobSpec is the technical blueprint for a Job.
type JobSpec struct {
	// PrimaryContainer specifies the details of an OCI container that forms the
	// cornerstone of the Job. Job success or failure is tied to completion and
	// exit code of this container.
	PrimaryContainer JobContainerSpec `json:"primaryContainer"`
	// SidecarContainers specifies the details of supplemental, "sidecar"
	// containers. Their completion and exit code do not directly impact Job
	// status. Brigade does not understand dependencies between a Job's multiple
	// containers and cannot enforce any specific startup or shutdown order. When
	// such dependencies exist (for instance, a primary container than cannot
	// proceed with a suite of tests until a database is launched and READY in a
	// sidecar container), then logic within those containers must account for
	// these constraints.
	SidecarContainers map[string]JobContainerSpec `json:"sidecarContainers,omitempty"` // nolint: lll
	// TimeoutSeconds specifies the time, in seconds, that must elapse before a
	// running Job should be considered to have timed out.
	TimeoutSeconds int64 `json:"timeoutSeconds,omitempty"`
	// Host specifies criteria for selecting a suitable host (substrate node) for
	// the Job. This is useful in cases where a Job requires a specific,
	// non-default operating system (i.e. Windows) or specific hardware (e.g. a
	// GPU.)
	Host *JobHost `json:"host,omitempty"`
}

// JobContainerSpec amends the ContainerSpec type with additional Job-specific
// fields.
type JobContainerSpec struct {
	// ContainerSpec encapsulates generic specifications for an OCI container.
	ContainerSpec `json:",inline"`
	// UseWorkspace indicates whether the Job requires the Worker's shared
	// workspace (if one exists) to be mounted into the OCI container.
	UseWorkspace bool `json:"useWorkspace"`
	// WorkspaceMountPath specifies the path in the OCI container's file system
	// where, if applicable, the Worker's shared workspace should be mounted.
	WorkspaceMountPath string `json:"workspaceMountPath,omitempty"`
	// UseSource indicates whether the Job requires source code to be retrieved
	// from a git repository and mounted into the OCI container. Note this
	// requires git configuration to have been specified at the Project and/or
	// Event levels.
	UseSource bool `json:"useSource"`
	// SourceMountPath specifies the path in the OCI container's file system
	// where, if applicable, source code retrieved from a git repository should be
	// mounted.
	SourceMountPath string `json:"sourceMountPath,omitempty"`
	// Privileged indicates whether the OCI container should operate in a
	// "privileged" (relaxed permissions) mode. This is commonly used to effect
	// "Docker-in-Docker" ("DinD") scenarios wherein one of a Job's OCI containers
	// must run its own Docker daemon. Note this field REQUESTS privileged status
	// for the container, but that may be disallowed by Project-level
	// configuration.
	Privileged bool `json:"privileged"`
	// UseHostDockerSocket indicates whether the OCI container should mount the
	// host's Docker socket into its own file system. This is commonly used to
	// effect "Docker-out-of-Docker" ("DooD") scenarios wherein one of a Job's OCI
	// containers must utilize the host's Docker daemon. GENERALLY, THIS IS HIGHLY
	// DISCOURAGED. Note this field REQUESTS to mount the host's Docker socket
	// into the container, but that may be disallowed by Project-level
	// configuration.
	UseHostDockerSocket bool `json:"useHostDockerSocket"`
}

// JobHost represents criteria for selecting a suitable host (substrate node)
// for a Job.
type JobHost struct {
	// OS specifies which "family" of operating system is required on a substrate
	// node to host a Job. Valid values are "linux" and "windows". When empty,
	// Brigade assumes "linux".
	OS string `json:"os,omitempty"`
	// NodeSelector specifies labels that must be present on the substrate node to
	// host a Job. This provides an opaque mechanism for communicating Job needs
	// such as specific hardware like an SSD or GPU.
	NodeSelector map[string]string `json:"nodeSelector,omitempty"`
}

// JobStatus represents the status of a Job.
type JobStatus struct {
	// Started indicates the time the Job began execution.
	Started *time.Time `json:"started,omitempty"`
	// Ended indicates the time the Job concluded execution. It will be nil
	// for a Job that is not done executing.
	Ended *time.Time `json:"ended,omitempty"`
	// Phase indicates where the Job is in its lifecycle.
	Phase JobPhase `json:"phase,omitempty"`
}

// MarshalJSON amends JobStatus instances with type metadata so that clients do
// not need to be concerned with the tedium of doing so.
func (j JobStatus) MarshalJSON() ([]byte, error) {
	type Alias JobStatus
	return json.Marshal(
		struct {
			meta.TypeMeta `json:",inline"`
			Alias         `json:",inline"`
		}{
			TypeMeta: meta.TypeMeta{
				APIVersion: meta.APIVersion,
				Kind:       "JobStatus",
			},
			Alias: (Alias)(j),
		},
	)
}

// JobsClient is the specialized client for managing Event Jobs with the
// Brigade API.
type JobsClient interface {
	// Create, given an Event identifier and Job, creates a new pending Job
	// and schedules it for execution.
	Create(
		ctx context.Context,
		eventID string,
		jobName string,
		job Job,
	) error
	// Start initiates execution of a pending Job.
	Start(
		ctx context.Context,
		eventID string,
		jobName string,
	) error
	// GetStatus, given an Event identifier and Job name, returns the Job's
	// status.
	GetStatus(
		ctx context.Context,
		eventID string,
		jobName string,
	) (JobStatus, error)
	// WatchStatus, given an Event identifier and Job name, returns a channel
	// over which the Job's status is streamed. The channel receives a new
	// JobStatus every time there is any change in that status.
	WatchStatus(
		ctx context.Context,
		eventID string,
		jobName string,
	) (<-chan JobStatus, <-chan error, error)
	// UpdateStatus, given an Event identifier and Job name, updates the status
	// of that Job.
	UpdateStatus(
		ctx context.Context,
		eventID string,
		jobName string,
		status JobStatus,
	) error
}

type jobsClient struct {
	*rm.BaseClient
}

// NewJobsClient returns a specialized client for managing Event Jobs.
func NewJobsClient(
	apiAddress string,
	apiToken string,
	opts *restmachinery.APIClientOptions,
) JobsClient {
	return &jobsClient{
		BaseClient: rm.NewBaseClient(apiAddress, apiToken, opts),
	}
}

func (j *jobsClient) Create(
	ctx context.Context,
	eventID string,
	jobName string,
	job Job,
) error {
	return j.ExecuteRequest(
		ctx,
		rm.OutboundRequest{
			Method:      http.MethodPut,
			Path:        fmt.Sprintf("v2/events/%s/worker/jobs/%s", eventID, jobName),
			AuthHeaders: j.BearerTokenAuthHeaders(),
			ReqBodyObj:  job,
			SuccessCode: http.StatusCreated,
		},
	)
}

func (j *jobsClient) Start(
	ctx context.Context,
	eventID string,
	jobName string,
) error {
	return j.ExecuteRequest(
		ctx,
		rm.OutboundRequest{
			Method: http.MethodPut,
			Path: fmt.Sprintf(
				"v2/events/%s/worker/jobs/%s/start",
				eventID,
				jobName,
			),
			AuthHeaders: j.BearerTokenAuthHeaders(),
			SuccessCode: http.StatusOK,
		},
	)
}

func (j *jobsClient) GetStatus(
	ctx context.Context,
	eventID string,
	jobName string,
) (JobStatus, error) {
	status := JobStatus{}
	return status, j.ExecuteRequest(
		ctx,
		rm.OutboundRequest{
			Method: http.MethodGet,
			Path: fmt.Sprintf(
				"v2/events/%s/worker/jobs/%s/status",
				eventID,
				jobName,
			),
			AuthHeaders: j.BearerTokenAuthHeaders(),
			SuccessCode: http.StatusOK,
			RespObj:     &status,
		},
	)
}

func (j *jobsClient) WatchStatus(
	ctx context.Context,
	eventID string,
	jobName string,
) (<-chan JobStatus, <-chan error, error) {
	resp, err := j.SubmitRequest(
		ctx,
		rm.OutboundRequest{
			Method: http.MethodGet,
			Path: fmt.Sprintf(
				"v2/events/%s/worker/jobs/%s/status",
				eventID,
				jobName,
			),
			AuthHeaders: j.BearerTokenAuthHeaders(),
			QueryParams: map[string]string{
				"watch": "true",
			},
			SuccessCode: http.StatusOK,
		},
	)
	if err != nil {
		return nil, nil, err
	}

	statusCh := make(chan JobStatus)
	errCh := make(chan error)

	go j.receiveStatusStream(ctx, resp.Body, statusCh, errCh)

	return statusCh, errCh, nil
}

func (j *jobsClient) UpdateStatus(
	ctx context.Context,
	eventID string,
	jobName string,
	status JobStatus,
) error {
	return j.ExecuteRequest(
		ctx,
		rm.OutboundRequest{
			Method: http.MethodPut,
			Path: fmt.Sprintf(
				"v2/events/%s/worker/jobs/%s/status",
				eventID,
				jobName,
			),
			AuthHeaders: j.BearerTokenAuthHeaders(),
			ReqBodyObj:  status,
			SuccessCode: http.StatusOK,
		},
	)
}

func (j *jobsClient) receiveStatusStream(
	ctx context.Context,
	reader io.ReadCloser,
	statusCh chan<- JobStatus,
	errCh chan<- error,
) {
	defer reader.Close()
	decoder := json.NewDecoder(reader)
	for {
		status := JobStatus{}
		if err := decoder.Decode(&status); err != nil {
			select {
			case errCh <- err:
			case <-ctx.Done():
			}
			return
		}
		select {
		case statusCh <- status:
		case <-ctx.Done():
			return
		}
	}
}

File: v2/brigadier-polyfill/src/jobs.ts
import { Logger } from "winston" 

import { Event, Job as BrigadierJob } from "@brigadecore/brigadier"

import { core } from "@brigadecore/brigade-sdk"

import { logger } from "./logger"

export class Job extends BrigadierJob {
  logger: Logger

  constructor(name: string, image: string, event: Event) {
    super(name, image, event)
    this.logger = logger.child({ job: name })
  }

  async run(): Promise<void> {
    this.logger.info(`Creating job ${this.name}`)
    try {
      const jobsClient = new core.JobsClient(
        this.event.worker.apiAddress,
        this.event.worker.apiToken,
        {allowInsecureConnections: true},
      )

      const sdkJob: core.Job = {
        name: this.name,
        spec: {
          primaryContainer: this.primaryContainer,
          sidecarContainers: this.sidecarContainers,
          timeoutDuration: this.timeoutSeconds + "s",
          host: this.host
        }
      }
      await jobsClient.create(this.event.id, sdkJob)
    }
    catch(e) {
      throw new Error(`Error creating job "${this.name}": ${e.message}`)
    }
    return this.wait()
  }

  private async wait(): Promise<void> {
    return new Promise<void>((resolve, reject) => {
      const jobsClient = new core.JobsClient(
        this.event.worker.apiAddress,
        this.event.worker.apiToken,
        {allowInsecureConnections: true},
      )

      const statusStream = jobsClient.watchStatus(this.event.id, this.name)
      statusStream.onData((status: core.JobStatus) => {
        this.logger.debug(`Current job phase is ${status.phase}`)
        if (!this.fallible) {
          switch (status.phase) {
          case core.JobPhase.Aborted:
            reject(new Error(`Job "${this.name}" was aborted`))
            break
          case core.JobPhase.Canceled:
            reject(new Error(`Job "${this.name}" was canceled before starting`))
            break
          case core.JobPhase.Failed:
            reject(new Error(`Job "${this.name}" failed`))
            break
          case core.JobPhase.SchedulingFailed:
            reject(new Error(`Job "${this.name}" scheduling failed`))
            break
          case core.JobPhase.Succeeded:
            resolve()
            break
          case core.JobPhase.TimedOut:
            reject(new Error(`Job "${this.name}" timed out`))
            break
          }
        }
      })
      statusStream.onReconnecting(() => {
        this.logger.warn("status stream connecting")
      })
      statusStream.onClosed(() => {
        const msg = "status stream closed"
        if (this.fallible) {
          this.logger.warn(msg)
          resolve()
        } else {
          reject(new Error(msg)) 
        }
      })
      statusStream.onError((e: Error) => {
        const msg = `Error watching status for job "${this.name}": ${e.message}`
        if (this.fallible) {
          this.logger.warn(msg)
          resolve()
        } else {
          reject(new Error(msg))
        }
      })
      statusStream.onDone(() => {
        resolve()
      })
    })
  }

  async logs(): Promise<string> {
    return new Promise<string>((resolve, reject) => {
      const logsClient = new core.LogsClient(
        this.event.worker.apiAddress,
        this.event.worker.apiToken,
        {allowInsecureConnections: true},
      )

      const logsStream = logsClient.stream(
        this.event.id,
        {job: this.name},
        {follow: false},
      )
      let logs = ""
      logsStream.onData((logEntry: core.LogEntry) => {
        if (logs != "") {
          logs += "\n"
        }
        if (logEntry.time) {
          logs += logEntry.time + ": "
        }
        logs += logEntry.message
      })
      logsStream.onReconnecting(() => {
        this.logger.warn("log stream connecting")
      })
      logsStream.onClosed(() => {
        reject("log stream closed")
      })
      logsStream.onError((e: Error) => {
        reject(new Error(`Error retrieving logs for job "${this.name}": ${e.message}`))
      })
      logsStream.onDone(() => {
        resolve(logs)
      })
    })
  }
}
